patroni/ctl.py

```diff
@@ -1372,3 +1373,222 @@ def format_pg_version(version):
+
+    If DSN is provided, gather all the available non-internal GUC values having configuration file,
+    postmaster command line or environment variable as a source and store them in the appropriate part
+    of Patroni configuration (``postgresql.parameters`` or ``bootsrtap.dcs.postgresql.parameters``).
+
+    :param scope: Scope parameter value to write into the configuration.
+    :param file: Full path to the configuration file to be created (/tmp/patroni.yml by default).
+    :param dsn: Optional dsn string for the local instance to get GUC values from.
+    :param bin_dir: Optional path to Postgres binaries. Prefered way to get the PG version.
+    """
+    from patroni.config import Config
+    from patroni.validator import get_major_version
+
+    no_value_msg = '#FIXME'
+    pg_version = None
+
+    dynamic_config = Config.get_default_config()
+    dynamic_config['postgresql']['parameters'] = dict(dynamic_config['postgresql']['parameters'])
+    config = {
+        'scope': scope,
+        'name': socket.gethostname(),
+        'bootstrap': {
+            'dcs': dynamic_config
+        },
+        'postgresql': {
+            'parameters': None,
+            'connect_address': no_value_msg,
+            'listen': no_value_msg,
+            'bin_dir': ''
+        },
+    }
+    if bin_dir:
+        config['postgresql']['bin_dir'] = bin_dir
+
+    if dsn:
+        enrich_config_from_running_instance(dsn, config, no_value_msg)
+    else:  # no source instance
+        replicator = os.getenv('PATRONI_REPLICATION_USERNAME', 'replicator')
+        config['postgresql']['authentication'] = {
+            'superuser': {
+                'username': os.getenv('PATRONI_SUPERUSER_USERNAME', 'postgres'),
+                'password': no_value_msg
+            },
+            'replication': {
+                'username': replicator,
+                'password': no_value_msg
+            }
+        }
+        config['postgresql']['pg_hba'] = [
+            'host all all 0.0.0.0/0 md5',
+            f'host replication {replicator} 127.0.0.1/32 md5'
+        ]
+
+    if bin_dir:
+        # obtain version from the binary
+        try:
+            pg_version = postgres_major_version_to_int(get_major_version(bin_dir))
+        except PatroniException as e:
+            raise PatroniCtlException(str(e))
+    elif dsn:
+        # obtain PostgreSQL version of the running instance if bin_dir is not provided
+        try:
+            with open(f"{config['postgresql']['data_dir']}/PG_VERSION") as f:
+                pg_version = f.readline()
+                if not pg_version:
+                    raise PatroniCtlException('Failed to obtain PostgreSQL version from the PG_VERSION file')
+                pg_version = postgres_major_version_to_int(pg_version.strip())
+        except OSError as e:
+            raise PatroniCtlException(f'Error reading PG_VERSION file: {e}')
```

Comment:
Hmm, should this branch be reachable? If there is a running process the `enrich_config_from_running_instance` will already fill `postgresql.bin_dir`

Reply:
can there be a hypothetical situation that `PG_VERSION` file is removed?

Reply:
Maybe, but I think postgres will not like it, because it checks it on start.

Reply:
🤦‍♀️ now I understand which code branch you mean. Yes, definitely redundant. Thanks!